syntax = "proto3";

package hourai.actions;

import "hourai/data/models.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";

enum ActionStatusCode {
  SUCCESS = 0;
  ERROR = 1;
  PENDING = 2;
  QUEUED = 3;
  STARTED = 4;
}

// The execution status of a given action
message ActionStatus {
  ActionStatusCode code = 1;
  string error_message = 2;
}

// Additional metadata regarding a given action
message ActionContext {
  // The human-readable description of the action.
  string description = 1;
  // The human-provided reason for the action.
  string reason = 2;
  // The duration of the action, after which it the action is undone
  google.protobuf.Duration duration = 3;
}

// The origin of the action taken by the bot
message ActionSource {
  repeated uint64 authorizer_id = 1;
  oneof source {
    hourai.models.BotCommand bot_command = 3;
  }
}

// The target of the action taken by the bot
message ActionTarget {
  repeated uint64 guild_id = 1;
  repeated uint64 user_id = 2;
  repeated uint64 channel_id = 3;
  repeated uint64 role_id = 4;
}

message ActionId {
  // Unix timestamp the action took place
  google.protobuf.Timestamp timestamp = 2;
  // A randomly generated fprint
  uint64 fprint = 2;
}

message Action {
  ActionId id = 1;
  ActionSource source = 2;
  ActionTarget target = 3;
  ActionContext context = 4;

  oneof action {
    SendMessage send_message = 5;
    KickMember kick_member = 6;
    BanMember ban_member = 7;
    ChangeMemberRole change_member_role = 8;
    RunBotCommand run_bot_command = 9;
  }
}

// Actions
message SendMessage {
  string content = 2;
}

message KickMember {
}

message BanMember {
}

message ChangeMemberRole {
  enum Type {
    TOGGLE = 0;
    ADD = 1;
    REMOVE = 2;
  }
  Type type = 1;
}

message RunBotCommand {
  string command = 1;
}
